common/predictive-text/unit_tests/headless/promise-store.js

```diff
@@ -0,0 +1,76 @@
+var assert = require('chai').assert;
+var sinon = require('sinon');
+
+let PromiseStore = require('../../build').PromiseStore;
+
+describe('PromiseStore', function () {
+  describe('.make()', function () {
+    it("should track a promise's callbacks", function () {
+      var promises = new PromiseStore();
+      // There should be no tracked promises.
+      assert.lengthOf(promises, 0);
+
+      // Add one promise.
+      new Promise(function (resolve, reject) {
+        promises.make(randomToken(), resolve, reject);
+      });
+      assert.lengthOf(promises, 1);
+    });
+
+    // TODO: test existing token
+  });
+
+  describe('.keep()', function () {
+    it('should call the resolve() function', function () {
+      var promises = new PromiseStore();
+      var token = randomToken();
+      var promise = new Promise(function (resolve, reject) {
+        promises.make(token, resolve, reject);
+      });
+
+      // Resolve the promise asynchronously.
+      var randomPayload = randomToken();
+      doLater(function () {
+        assert.lengthOf(promises, 1);
+        promises.keep(token, randomPayload);
+      });
+
+      return promise.then(function (actual) {
```

Comment:
About this.  I'm fairly certain that promise.then merely returns a `Promise` of its own, regardless of whether or not it's been fulfilled - in other words, calling `.then()` is not blocking except within its internal block.  You'll want something more like the following:

```
it('should call the resolve() function', function (done) {
  // ...
  return promise.then(function (actual) {
    assert.strictEqual(actual, randomPayload);
    assert.lengthOf(promises, 0);
    done();
  });
});
```

The optional `done()` function parameter provided by Mocha provides an asynchronous way to signal completion of a test, which is needed in these scenarios.  If you wish to confirm that this is needed, simply try auto-failing one of these promise-based tests within the `.then()` section and watch what occurs.

Reply:
I'm taking advantage of Mocha's support for promises.

From [Working with Promises](https://mochajs.org/#working-with-promises):

> Alternately, instead of using the done() callback, you may return a Promise. This is useful if the APIs you are testing return promises instead of taking callbacks:
>
>     beforeEach(function() {
>       return db.clear()
>         .then(function() {
>           return db.save([tobi, loki, jane]);
>         });
>     });

Because `.then()` returns a promise, Mocha is able to wait for it, asynchronously, and obviates the need to call the `done()` callback.

Reply:
Oh! And if you do provide both:

```javascript
    it('...', function (done) {
      return promise.then(function (actual) {
        assert.strictEqual(actual, randomPayload);
        assert.lengthOf(promises, 0);
        done();
      });
    });
```

Mocha fails the test with this message:

```
  18 passing (55ms)
  1 failing

  1) PromiseStore
       .keep()
         should call the resolve() function:
     Error: Resolution method is overspecified. Specify a callback *or* return a Promise; not both.
```

Reply:
Ah, that sounds fine then.  My mistake.